### ISSUE TITLE
NMNIST CNN not generalising on full NMNIST

### ISSUE THREAD

**Issue by Afstkla** [bamsumit/slayerPytorch#56]
Hi there,

I'm running a clean clone of this repo, with the only changes being that I've switched from the small to the full NMNIST dataset. I've noticed previously I forgot to shuffle the samples, which resulted in horrible accuracy both in training and testing, but now, even though both train and test data are shuffled, I still get around 10% testing accuracy. I understand that the network is different from the network reported in the paper, but I would expect it to have at least somewhat decent testing accuracy.

After 9 epochs of training I'm at a training accuracy of 87%, and a validation accuracy of 10%.

I unfortunately don't have a public fork of the exact code, but this is the patch for `nmnistCNN.py`:
```python
diff --git a/example/03_NMNIST_CNN/nmnistCNN.py b/example/03_NMNIST_CNN/nmnistCNN.py
index 5605400..c838df7 100644
--- a/example/03_NMNIST_CNN/nmnistCNN.py
+++ b/example/03_NMNIST_CNN/nmnistCNN.py
@@ -4,6 +4,8 @@ sys.path.append(CURRENT_TEST_DIR + "/../../src")
 
 from datetime import datetime
 import numpy as np
+import matplotlib
+matplotlib.use('agg')
 import matplotlib.pyplot as plt
 import torch
 from torch.utils.data import Dataset, DataLoader
@@ -26,7 +28,7 @@ class nmnistDataset(Dataset):
         classLabel  = self.samples[index, 1]
         
         inputSpikes = snn.io.read2Dspikes(
-                        self.path + str(inputIndex.item()) + '.bs2'
+                        self.path + str(inputIndex.item()).zfill(5) + '.bin'
                         ).toSpikeTensor(torch.zeros((2,34,34,self.nTimeBins)),
                         samplingTime=self.samplingTime)
         desiredClass = torch.zeros((10, 1, 1, 1))
@@ -75,9 +77,9 @@ if __name__ == '__main__':
     # deviceIds = [2, 3, 1]
 
     # Create network instance.
-    net = Network(netParams).to(device)
+    # net = Network(netParams).to(device)
     # Split the network to run over multiple GPUs
-    # net = torch.nn.DataParallel(Network(netParams).to(device), device_ids=deviceIds)
+    net = torch.nn.DataParallel(Network(netParams).to(device))
 
     # Create snn loss instance.
     error = snn.loss(netParams).to(device)
@@ -90,13 +92,13 @@ if __name__ == '__main__':
                                 sampleFile  =netParams['training']['path']['train'],
                                 samplingTime=netParams['simulation']['Ts'],
                                 sampleLength=netParams['simulation']['tSample'])
-    trainLoader = DataLoader(dataset=trainingSet, batch_size=12, shuffle=False, num_workers=4)
+    trainLoader = DataLoader(dataset=trainingSet, batch_size=80, shuffle=True, num_workers=4)
 
     testingSet = nmnistDataset(datasetPath  =netParams['training']['path']['in'], 
                                 sampleFile  =netParams['training']['path']['test'],
                                 samplingTime=netParams['simulation']['Ts'],
                                 sampleLength=netParams['simulation']['tSample'])
-    testLoader = DataLoader(dataset=testingSet, batch_size=12, shuffle=False, num_workers=4)
+    testLoader = DataLoader(dataset=testingSet, batch_size=80, shuffle=True, num_workers=4)
 
     # Learning stats instance.
     stats = learningStats()
@@ -167,6 +169,8 @@ if __name__ == '__main__':
     plt.ylabel('Loss')
     plt.legend()
 
+    plt.savefig('LossLog.svg')
+
     plt.figure(2)
     plt.plot(stats.training.accuracyLog, label='Training')
     plt.plot(stats.testing .accuracyLog, label='Testing')
@@ -174,4 +178,6 @@ if __name__ == '__main__':
     plt.ylabel('Accuracy')
     plt.legend()
 
+    plt.savefig('AccuracyLog.svg')
+
     plt.show()
```

I've used the default NMNIST dataset from https://www.garrickorchard.com/datasets/n-mnist and formatted the test.txt and train.txt files just like the NMNISTsmall example (left NMNIST large, right NMNIST small):
![image](https://user-images.githubusercontent.com/9075380/113283431-9aac8e80-92e8-11eb-8229-c69da8c99a34.png)

Does anyone by any chance have an idea what could be going wrong?

Thanks in advance

**Comment by Afstkla** [comment]
Nevermind, apparently my test and train data overwrote each other, so I was validating incorrectly...
